### PR TITLE
[Spring] Make symmetric Request and Response API

### PR DESCRIPTION
**Pull request by AlexAPB**
Each commit is individually reviewable.

This change
- Removes `ResponseWriter` in favour of using the `Response` struct directly
- Removes `RequestMeta` in favour of using the `Request` struct directly
- Introduces a positional `yarpc.Buffer` (will need later modification) instead
  of embedding into the `Request` and `Response`s

This means `Request` and `Response` hold meta information about a corresponding
`yarpc.Buffer` payload.

Next up will be modifying `yarpcjson` and `yarpchttp` to match the API change.
Note: Build errors are expected as this is being merged into a feature branch.

### CODE REVIEW

File: etc/bin/generate.sh [hunk 1]
```diff
@@ -78,6 +79,11 @@ mockgen -destination=api/transport/transporttest/router.go -package=transporttes
 mockgen -destination=api/transport/transporttest/stream.go -package=transporttest go.uber.org/yarpc/api/transport Stream,StreamCloser
 mockgen -destination=api/transport/transporttest/transport.go -package=transporttest go.uber.org/yarpc/api/transport Transport
 
+# v2 mocks
+mockgen -destination=v2/yarpcmiddlewaretest/mock_router.go -package=yarpcmiddlewaretest go.uber.org/yarpc/v2/yarpcmiddleware Router,UnaryInbound,UnaryOutbound,StreamInbound,StreamOutbound
```
Comment (amckinney) [inline etc/bin/generate.sh]: We should consider merging `yarpcmiddlewaretest` into `yarpctest` later on.

File: v2/yarpcmiddleware/inbound.go [hunk 2]
```diff
@@ -55,26 +53,26 @@ func ApplyUnaryInbound(h yarpc.UnaryHandler, i UnaryInbound) yarpc.UnaryHandler
 }
 
 // UnaryInboundFunc adapts a function into an InboundMiddleware.
-type UnaryInboundFunc func(context.Context, *yarpc.Request, yarpc.ResponseWriter, yarpc.UnaryHandler) error
+type UnaryInboundFunc func(context.Context, *yarpc.Request, yarpc.UnaryHandler) (*yarpc.Response, error)
 
 // Handle for UnaryInboundFunc
-func (f UnaryInboundFunc) Handle(ctx context.Context, req *yarpc.Request, resw yarpc.ResponseWriter, h yarpc.UnaryHandler) error {
-	return f(ctx, req, resw, h)
+func (f UnaryInboundFunc) Handle(ctx context.Context, req *yarpc.Request, res, h yarpc.UnaryHandler) (*yarpc.Response, error) {
```
Comment (amckinney) [inline v2/yarpcmiddleware/inbound.go]: We need to remove `res` here.

File: v2/inbound_call.go [hunk 3]
```diff
@@ -81,21 +81,12 @@ func (ic *InboundCall) ReadFromRequest(req *Request) error {
...
-	return nil
-}
-
 // WriteToResponse writes response information from the InboundCall onto the
 // given Response.
 //
 // If used, this must be called before writing the response body to the
 // Response.
-func (ic *InboundCall) WriteToResponse(res *Request) {
+func (ic *InboundCall) WriteToResponse(res *Response) {
```
Comment (amckinney) [inline v2/inbound_call.go]: Was this just blatantly wrong before?

File: v2/yarpctest/fake_outbound.go [hunk 4]
```diff
@@ -66,6 +66,9 @@ func (t *FakeTransport) NewOutbound(c yarpc.Chooser, opts ...FakeOutboundOption)
 	return o
 }
 
+var _ yarpc.UnaryOutbound = (*FakeOutbound)(nil)
+var _ yarpc.StreamOutbound = (*FakeOutbound)(nil)
```
Comment (amckinney) [inline v2/yarpctest/fake_outbound.go]: Nice :+1:

File: v2/yarpctransport/handler_test.go [hunk 5]
```diff
@@ -25,36 +25,37 @@ import (
...
 
-type UnaryHandlerFunc func(context.Context, *Request) (*Response, error)
-type StreamHandlerFunc func(*ServerStream) error
+type UnaryHandlerFunc func(context.Context, *yarpc.Request, *yarpc.Buffer) (*yarpc.Response, *yarpc.Buffer, error)
+type StreamHandlerFunc func(*yarpc.ServerStream) error
 
-func (f UnaryHandlerFunc) Handle(ctx context.Context, r *Request) (*Response, error) {
-	return f(ctx, r, w)
+func (f UnaryHandlerFunc) Handle(ctx context.Context, r *yarpc.Request, b *yarpc.Buffer) (*yarpc.Response, *yarpc.Buffer, error) {
+	return f(ctx, r, b)
```
Comment (amckinney) [inline v2/yarpctransport/handler_test.go]: Curious why there isn't syntax highlighting here. I wonder if there is a compilation issue somewhere?

File: v2/inbound_call.go (hunk 3 above)
Comment (AlexAPB) [reply v2/inbound_call.go]: my mistake, this was changed incorrectly in the previous commit

File: v2/yarpctransport/handler_test.go (hunk 5 above)
Comment (AlexAPB) [reply v2/yarpctransport/handler_test.go]: another test package is importing `yarpctest` which still has some work needed, I think that's what's affecting this file.

**Comment by AlexAPB** [comment]
Fixed your comments @amckinney

**Pull request merged**